### PR TITLE
Show gain and loss in one color scale legend item

### PR DESCRIPTION
**Pull request by creilly8**
## Description
Closes issue #2374. 

The cnv gain and loss legend item data is merged into one legend item in `svg.legend`. The goal was to recreate the cnv [color scale in the mds3 track.](http://localhost:3000/?genome=hg38&gene=p16&mds3=ASH) 

Changes: 
1. Increase code encapsulation for the color scale, code reorg
2. Update svg.legend to combine gain and loss legend items into one a parents property
3. Minor update to the disco code to support the merged gain and loss items
4. Minor updates to the matrix interactivity code to act on parents of a legend item to maintain the functionality. 

Test
1. [Disco example](http://localhost:3000/?genome=hg38&dslabel=GDC&disco=1&sample=TCGA-AP-A059-01A)
5. [Matrix example](http://localhost:3000/?noheader=1&mass=%7B%22dslabel%22:%22ASH%22,%22genome%22:%22hg38%22,%22plots%22:%5B%7B%22chartType%22:%22matrix%22,%22termgroups%22:%5B%7B%22lst%22:%5B%7B%22id%22:%22Sex%22%7D,%7B%22term%22:%7B%22gene%22:%22TAL1%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22KRAS%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22MYB%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22PAX5%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22NOTCH1%22,%22type%22:%22geneVariant%22%7D%7D%5D%7D%5D%7D%5D%7D)

## Checklist
[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

**Comment by xzhou82** [comment]
[ash matrix](http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22ASH%22,%22genome%22:%22hg38%22,%22plots%22:[{%22chartType%22:%22matrix%22,%22termgroups%22:[{%22lst%22:[{%22term%22:{%22gene%22:%22MYC%22,%22type%22:%22geneVariant%22}},{%22term%22:{%22gene%22:%22CDKN2A%22,%22type%22:%22geneVariant%22}},{%22term%22:{%22gene%22:%22NOTCH1%22,%22type%22:%22geneVariant%22}}]}]}]}) cnv legend doesn't show properly

**Review by xzhou82 — APPROVED** [review]
works for me but @congyu-lu or @siosonel should review matrix changes, and if to work on matrix cnv fix in a subsequent branch (use consistent scale for gain and loss, use consistent colors for legend and cell rendering, click cnv legend to trigger native menu with options auto/fixed/percentile

File: client/plots/matrix/matrix.legend.js [hunk 1]
```diff
@@ -143,6 +143,54 @@ export function getLegendData(legendGroups, refs, self) {
 		const keys = Object.keys(legend.values).sort()
 		const hasScale = Object.values(legend.values).find(v => v.scale)
 		if (hasScale) {
+			if (legend === legendGroups?.CNV) {
+				/** Combine gain and loss here to avoid rendering issues */
+				const legendValues = Object.values(legend.values)
+				const findLegend = type =>
```
Comment (siosonel) [inline client/plots/matrix/matrix.legend.js:149]: should use `v.key === 'CNV_loss'` (or use constant value from `common.js`) instead of comparing against `v.label` which is arbitrary.

File: client/plots/matrix/matrix.legend.js [hunk 2]
```diff
@@ -143,6 +143,54 @@ export function getLegendData(legendGroups, refs, self) {
...
+					legendValues.find(v => v.label.toLowerCase().includes(type) || v.key.toLowerCase().includes(type))
+
+				const gain = findLegend('gain')
+				const loss = findLegend('loss')
+
+				if (gain?.scale && loss?.scale) {
+					const gainColors = getColors(gain)
+					const lossColors = getColors(loss)
+					const colors = [lossColors[1], 'white', gainColors[1]]
+					const domain = setColorScaleDomain(loss.maxLabel, gain.maxLabel, [0, 0], colors)
```
Comment (siosonel) [inline client/plots/matrix/matrix.legend.js:159]: should use `loss.domain[1]` and `gain.domain[1]`, the actual numeric values instead of extracting presumed numeric values from string labels

File: client/dom/svg.legend.js [hunk 3]
```diff
@@ -189,21 +189,30 @@ export default function svgLegend(opts) {
 		let colorGradientId
 		if (d.domain) {
 			colorGradientId = `sjpp-linear-gradient-${getId()}`
-			new ColorScale({
+			const data = d.domain || [d.minLabel, d.maxLabel]
```
Comment (siosonel) [inline client/dom/svg.legend.js:192]: If `data` is used to compute scales or other values, then the defaults of using `minLabel` and `maxLabel` for `data` is risky, since a label can contain string like `0.1 mg`. Should use a data property that is guaranteed to be raw numeric values.

File: client/plots/matrix/matrix.legend.js (hunk 2 above)
Comment (creilly8) [reply client/plots/matrix/matrix.legend.js:159]: @siosonel, the problem with using the domain for the cnv values is they are often not the real data. Usually the domains were set to [0,1]. Only the min and max label contained the real data.

Comment (siosonel) [reply client/plots/matrix/matrix.legend.js:159]: Based on `matrix.cell.js` (see screenshot below), `loss.maxLabel` is equal to `loss.domain[1]` or `-minLoss`, and `gain.maxLabel === gain.domain[1]` or `maxGain`, and these refer the actual numeric values. Whereas variable names with `"label"` in it may eventually include units or comma-separators, these are not guaranteed to be fully numeric values in the future. And by using numbers that are taken directly from the `domain[1]`, which was used for the scale to color matrix CNV cells, it ensures that the legend colors match the rendered data value colors. But if there are other technical concerns about using domain, we can discuss.

Screenshot 2024-12-03 at 8 33 22 PM
 `setColorScaleDomain()`.

**Comment by siosonel** [comment]
@creilly8 Instead of having making a child scale `legend.values.CNV_gain_loss` from "parent" scales, it would be much simpler and safer for the matrix CNV cells to use only one color scale that spans negative, 0, positive values, and for the CNV legend to reuse that scale. This way, the legend will also be guaranteed to show the same color representation in the matrix cells.

File: client/plots/matrix/matrix.legend.js (hunk 2 above)
Comment (creilly8) [reply client/plots/matrix/matrix.legend.js:159]: That's still not the real data in every instance. `-minloss` converts the real minimum CNV loss value to a positive value in every instance. The maxLabel is the correct domain[1] for each item.

Comment (creilly8) [reply client/plots/matrix/matrix.legend.js:159]: I intentionally left the rendering code alone since how the scales are used are not the same.

File: client/plots/matrix/matrix.legend.js (hunk 1 above)
Comment (siosonel) [reply client/plots/matrix/matrix.legend.js:149]: removed this code in my clean up, so no need to adress

File: client/plots/matrix/matrix.legend.js (hunk 2 above)
Comment (siosonel) [reply client/plots/matrix/matrix.legend.js]: addressed in my fix

File: client/dom/svg.legend.js (hunk 3 above)
Comment (siosonel) [reply client/dom/svg.legend.js:192]: Addressed by using `minLoss`, `maxGain` in `matrix.layout` and supplying precomputed legend domain, range to `matrix.legend`

**Comment by siosonel** [comment]
The legend colors now more closely matches the rendered matrix cell colors, by including lots of interpolated domain and range entries using `getInterpolatedDomainRange()` helper. Previously, the reds were darker in the legend while the matrix cell was brighter/more orange in lower gain values.

Screenshot 2024-12-04 at 7 48 54 PM

**Comment by creilly8** [comment]
@xzhou82, @siosonel - I agree with Edgar's changes. I've tested these changes in all the other color scale implementations and do not see any problems specific to this PR.

**Review by congyu-lu — COMMENTED** [review]
Screenshot 2024-12-05 at 09 58 20

for the ticks on CNV scale, 0 is not in the middle for some reason

**Comment by creilly8** [comment]
> Screenshot 2024-12-05 at 09 58 20
> for the ticks on CNV scale, 0 is not in the middle for some reason [url](http://localhost:3000/?noheader=1&mass=%7B%22dslabel%22:%22ASH%22,%22genome%22:%22hg38%22,%22plots%22:%5B%7B%22chartType%22:%22matrix%22,%22termgroups%22:%5B%7B%22lst%22:%5B%7B%22id%22:%22Sex%22%7D,%7B%22term%22:%7B%22gene%22:%22TAL1%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22KRAS%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22MYB%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22MYC%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22CDKN2A%22,%22type%22:%22geneVariant%22%7D%7D,%7B%22term%22:%7B%22gene%22:%22NOTCH1%22,%22type%22:%22geneVariant%22%7D%7D%5D%7D%5D%7D%5D%7D)

@congyu-lu, This is fixed in https://github.com/stjude/proteinpaint/pull/2423/commits/212d26084f76e4e1313814a634357cae8c12e685. I rebased this branch to check all the changes against the new commits on master Please delete your local copy and check again.

**Review by congyu-lu — COMMENTED** [review]
pulled the latest but still saw 0 was not in the middle of the color scale

**Comment by creilly8** [comment]
> pulled the latest but still saw 0 was not in the middle of the color scale

Would you pull and try again?

**Comment by congyu-lu** [comment]
> url

thanks it works now, but if you click and select "hide samples with CNV gain", the color scale seems to have "0" and "-20" overlapped with each other

image

**Comment by creilly8** [comment]
> > url
> 
> thanks 0 is in the middle now, but if you click and select "hide samples with CNV gain", the color scale seems to have "0" and "-20" overlapped with each other
> 
> image

The majority of the cnv loss values for the domain are converted to a positive value from matrix.layout.

**Review by congyu-lu — COMMENTED** [review]
sorry the matrix color scale ticks still have issues. 
1. Clicking "hide samples with loss, then see overlapped "1" and "2". 
2. Clicking "Mutations" and then "do not show mutations", the color scale only has "0" left.

**Comment by congyu-lu** [comment]
> > sorry the matrix color scale ticks still have issues. I am ok if you want to merge it and fix them later
> > 
> > 1. Clicking "hide samples with loss, then see overlapped "1" and "2".
> > 2. Clicking "Mutations" and then "do not show mutations", the color scale only has "0" left.
> 
> Those errors are data related. For example, the 2 is only value out of 103 above 1, the outlier situation we discussed on Monday. I'd rather merge this and work on each one of these problems separately.

ok sounds good

**Review by congyu-lu — APPROVED** [review]

**Review by siosonel — APPROVED** [review]
